### PR TITLE
test: Add e2e test for streaming updates python client

### PR DESCRIPTION
**Pull request by gt2345**
## Description

Add e2e test for streaming updates python client

## Test Plan

e2e tests pass

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[MD-282](https://hpe-aiatscale.atlassian.net/browse/MD-282)

[MD-282]: https://hpe-aiatscale.atlassian.net/browse/MD-282?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: e2e_tests/pytest.ini [hunk 1]
```diff
@@ -13,6 +13,7 @@ markers =
     e2e_cpu_postgres: end to end CPU tests for testing basic database functionality
     e2e_cpu_cross_version: end to end CPU tests for testing basic cluster functionality with differing master/agent versions
     e2e_cpu_agent_connection_loss: end to end CPU tests for testing agent functionality on connection loss
+    e2e_cpu_streaming: end to end streaming updates tests
```
Comment (rb-determined-ai) [inline e2e_tests/pytest.ini:16]: why add a new kind of test?  Why not include this in the normal cpu test suite?

File: e2e_tests/tests/streaming/test_client.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import uuid
+
+import pytest
+
+from determined.common.api import bindings
+from determined.common.streams import wire
+from determined.common.streams._client import LomondStreamWebSocket, ProjectSpec, Stream, Sync
```
Comment (rb-determined-ai) [inline e2e_tests/tests/streaming/test_client.py:7]: don't import symbols directly; only import modules.

In this case, `from determined.common import streams` and use `streams.LomondStreamWebSocket`, `streams.ProjectSpec`, etc.

File: e2e_tests/tests/streaming/test_client.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    sess = api_utils.admin_session()
+    ws = LomondStreamWebSocket(sess)
+    stream = Stream(ws)
+
+    syncId = "sync1"
+    projectName = "streaming_project"
+    newProjectName = "streaming_project_1"
+
+    resp_w = bindings.post_PostWorkspace(
+        sess, body=bindings.v1PostWorkspaceRequest(name=f"workspace_aug_{uuid.uuid4().hex[:8]}")
```
Comment (rb-determined-ai) [inline e2e_tests/tests/streaming/test_client.py:39]: how about `codecs.encode(random.randbytes(4), 'hex')` instead?

Using `uuid.uuid4()` to get random bytes is odd.

File: e2e_tests/tests/streaming/test_client.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    bindings.delete_DeleteProject(sess, id=p.id)
+    for event in stream:
+        if isinstance(event, wire.ProjectMsg):
+            assert event.state == "DELETING"
+        elif isinstance(event, wire.ProjectsDeleted):
+            assert event == wire.ProjectsDeleted(str(p.id))
+            break
+        else:
+            raise Exception("Unexpected message from stream.")
```
Comment (rb-determined-ai) [inline e2e_tests/tests/streaming/test_client.py:76]: don't raise a bare Exception.  Use a `ValueError` if you have to do something whose meaning doesn't convey much.

Also, please include the message in the exception, so that if anybody were to hit the error in CI they'd see the message that caused the problem in its entirety.

File: e2e_tests/pytest.ini (hunk 1 above)
Comment (gt2345) [reply e2e_tests/pytest.ini:16]: I wonder what is the rule for splitting tests? I saw rbac, elastic being tested separately

File: e2e_tests/tests/streaming/test_client.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
+import random
+
+import pytest
+
+from determined.common import streams
+from determined.common.api import bindings
+from determined.common.streams import _client
```
Comment (rb-determined-ai) [inline e2e_tests/tests/streaming/test_client.py:7]: You are importing the modules, yes, but in the case of the `streams._client` module, there is nothing that calling code should be importing directly.

The full code layout rfc is [here](https://github.com/rb-determined-ai/pedl/blob/9bd2168ec23fbba866194f3d7d3cfc243d267100/python_code_layout.md) if you want to read it.

TL;DR is: you can generally assume that any of our python files that start with an underscore are not meant to be imported directly, and the relevant symbols should be available in their parent namespace.

So in this case, all the symbols you get from `_client.*` should actually be accessed as `stream.*`.

You can think of this as a way to separate the organization of source code (which code goes in which `_`-prefixed files) from the import path for accessing exported symbols.

**Review by rb-determined-ai — APPROVED** [review]
thank you!

**Review by rb-determined-ai — APPROVED** [review]
